### PR TITLE
[MI-3490] Added the logic of streaming files from MS Teams and uploading directly to MM without occupying memory

### LINKED ISSUE #297: We should stream large files from Teams to MM instead of loading the whole file in memory

### PR DESCRIPTION
**Pull request by manojmalik20**
#### Summary
- Updated the Mattermost server version being used in go.mod to 7.8.10
- Modified the existing client functions and created new functions to handle the usecase
- Created two new config settings for maxSizeForCompleteDownload and bufferSizeForStreaming
- Added the logic to download whole file into memory if file size is less than or equal to a new config setting
- Added the logic to stream file from MS Teams and upload to Mattermost in chunks of buffer size defined in the new config setting

#### Ticket Link
Fixes #297

### CODE REVIEW

**Review by ayusht2810 — APPROVED** [review]

File: plugin.json [hunk 1]
```diff
@@ -99,6 +99,20 @@
           "type": "number",
           "help_text": "Set the time for DMs and GMs after which the user will get a prompt to connect their account when they try to post a message without connecting their account. (Leave it empty to disable the prompt)",
           "default": 0
+        },{
+        },{
+          "key": "maxSizeForCompleteDownload",
+          "display_name": "Maximum size of attachments to support complete one time download (in MB)",
+          "type": "number",
+          "help_text": "Set the maximum size for attachments which can be loaded into the memory. Attachments bigger than this size will be streamed from MS Teams to Mattermost",
```
Comment (ayusht2810) [inline plugin.json:107]:
```suggestion
          "help_text": "Set the maximum size for attachments that can be loaded into the memory. Attachments bigger than this size will be streamed from MS Teams to Mattermost",
```

**Review by ayusht2810 — APPROVED** [review]

File: server/msteams/client.go [hunk 2]
```diff
@@ -1211,30 +1212,70 @@ func (tc *ClientImpl) GetFileContent(weburl string, fileSizeAllowed int64) ([]by
...
+	data, err := drives.NewItemItemsItemContentRequestBuilder(downloadURL, tc.client.RequestAdapter).Get(tc.ctx, nil)
 	if err != nil {
 		return nil, NormalizeGraphAPIError(err)
 	}
 	return data, nil
 }
 
+func (tc *ClientImpl) GetFileContentStream(downloadURL string, writer *io.PipeWriter, bufferSize int64) {
+	rangeStart := int64(0)
+	// Get 10 MB of data from the API call in one iteration
```
Comment (Kshitij-Katiyar) [inline server/msteams/client.go:1245]: remove this

File: server/msteams/client.go [hunk 3]
```diff
@@ -1211,30 +1212,70 @@ func (tc *ClientImpl) GetFileContent(weburl string, fileSizeAllowed int64) ([]by
...
+			return
+		}
+
+		if _, err = io.Copy(writer, res.Body); err != nil {
+			tc.logError("unable to copy response body to the writer", "error", err.Error())
+			return
+		}
+
+		res.Body.Close()
+		contentLenStr := res.Header.Get("Content-Length")
```
Comment (Kshitij-Katiyar) [inline server/msteams/client.go:1268]:
```suggestion
		contentLength := res.Header.Get("Content-Length")
```

File: server/handlers/attachments.go [hunk 4]
```diff
@@ -104,49 +131,58 @@ func (ah *ActivityHandler) handleAttachments(channelID, text string, msg *msteam
...
+				ah.plugin.GetAPI().LogError("failed to get file size and download URL", "error", err.Error())
 				continue
 			}
 
-			imageRes := int64(w) * int64(h)
-			if imageRes > *ah.plugin.GetAPI().GetConfig().FileSettings.MaxImageResolution {
-				ah.plugin.GetAPI().LogError("image resolution is too high")
+			fileSizeAllowed := *ah.plugin.GetAPI().GetConfig().FileSettings.MaxFileSize
+			if fileSize > fileSizeAllowed {
+				ah.plugin.GetAPI().LogError("skipping file download from MS Teams because the file size is greater than the allowed size")
```
[the remark this answers is not in the document]
Comment (manojmalik20) [reply server/handlers/attachments.go:153]: No

File: server/handlers/attachments.go [hunk 5]
```diff
@@ -104,49 +131,58 @@ func (ah *ActivityHandler) handleAttachments(channelID, text string, msg *msteam
...
-		}
 
-		if a.Name == "" {
-			extension := ""
-			extensions, extensionErr := mime.ExtensionsByType(contentType)
-			if extensionErr != nil {
-				ah.plugin.GetAPI().LogDebug("Unable to get the extensions using content type", "error", extensionErr.Error())
-			} else if len(extensions) > 0 {
-				extension = extensions[0]
+			// If the file size is less than or equal to 20 MB, then download the file directly instead of streaming
```
[the remark this answers is not in the document]
Comment (manojmalik20) [reply server/handlers/attachments.go:158]: Modified it

**Review by Kshitij-Katiyar — APPROVED** [review]

**Pull request merged**